apps/prairielearn/src/lib/externalGrader.js

```diff
@@ -1,58 +1,73 @@
-const ERR = require('async-stacktrace');
-const async = require('async');
-const { config } = require('./config');
-const { logger } = require('@prairielearn/logger');
-const sqldb = require('@prairielearn/postgres');
-const externalGradingSocket = require('./externalGradingSocket');
+// @ts-check
+import * as async from 'async';
+import * as util from 'util';
+import { z } from 'zod';
+import { logger } from '@prairielearn/logger';
+import * as sqldb from '@prairielearn/postgres';
+
+import { config } from './config';
+import * as externalGradingSocket from './externalGradingSocket';
 const ExternalGraderSqs = require('./externalGraderSqs');
 const ExternalGraderLocal = require('./externalGraderLocal');
-const assessment = require('./assessment');
+import * as assessment from './assessment';
+import {
+  CourseSchema,
+  QuestionSchema,
+  GradingJobSchema,
+  SubmissionSchema,
+  VariantSchema,
+} from './db-types';
 
 const sql = sqldb.loadSqlEquiv(__filename);
 
-module.exports.init = function (callback) {
+/** @typedef Grader
+ * @property {(GradingJob, Submission, Variant, Question, Course) => import('events')} handleGradingRequest
+ */
+
+/** @type {Grader | null} */
+let grader = null;
+
+export function init() {
   if (config.externalGradingUseAws) {
     logger.verbose('External grader running on AWS');
-    module.exports.grader = new ExternalGraderSqs();
-    callback(null);
+    grader = new ExternalGraderSqs();
   } else {
     // local dev mode
     logger.verbose('External grader running locally');
-    module.exports.grader = new ExternalGraderLocal();
-    callback(null);
+    grader = new ExternalGraderLocal();
   }
-};
-
-module.exports.beginGradingJob = function (grading_job_id, callback) {
-  const params = {
-    grading_job_id,
-  };
-  sqldb.queryOneRow(sql.select_grading_job_info, params, (err, result) => {
-    if (ERR(err, callback)) return;
-    const { grading_job, submission, variant, question, course } = result.rows[0];
-
-    module.exports._beginGradingJob(grading_job, submission, variant, question, course);
-    callback(null);
-  });
-};
-
-module.exports.beginGradingJobs = function (grading_job_ids, callback) {
-  async.each(
-    grading_job_ids,
-    (grading_job_id, callback) => {
-      module.exports.beginGradingJob(grading_job_id, (err) => {
-        if (ERR(err, callback)) return;
-        callback(null);
-      });
-    },
-    (err) => {
-      if (ERR(err, callback)) return;
-      callback(null);
-    },
+}
+
+/**
+ * @param {string[]} grading_job_ids
+ */
+export async function beginGradingJobsAsync(grading_job_ids) {
+  await async.each(grading_job_ids, beginGradingJobAsync);
+}
+export const beginGradingJobs = util.callbackify(beginGradingJobsAsync);
```

Comment:
AFAICT, the callback-based version is only used in one place:

https://github.com/PrairieLearn/PrairieLearn/blob/78b3ee3c7bfffb7f07127eafbdc7d8692c2a831d/apps/prairielearn/src/lib/assessment.js#L251-L261

This would be pretty trivial to refactor into a Promise-based call, if you wanted to eliminate `beginGradingJobs` here.

Reply:
Done in 95dc698f25f22b2264aeccc8476a22314852a296. I also removed the `Async` suffix. I've also done the same for `beginGradingJob`, but I kept the current calls as callbackify versions for now instead of refactoring the code.